src/core/init/initialize_media_source.ts

```diff
@@ -254,133 +257,156 @@ export default function InitializeOnMediaSource(
     .pipe(filter((evt) : evt is IManifestFetcherParsedResult => evt.type === "parsed"));
 
   /** Load and play the content asked. */
-  const loadContent$ = observableCombineLatest([initialManifest$,
-                                                openMediaSource$,
-                                                waitForEMEReady$]).pipe(
-    mergeMap(([parsedManifest, initialMediaSource]) => {
-      const manifest = parsedManifest.manifest;
+  const loadContent$ = emeManager$.pipe(
+    filter((evt) => evt.type === "created-media-keys" ||
+                    evt.type === "eme-disabled"),
+    take(1),
+    mergeMap((emeEvt) => {
+      return openMediaSource$.pipe(
+        mergeMap((initialMediaSource) => {
+          if (emeEvt.type === "created-media-keys") {
+            emeEvt.value.attachMediaKeys$.next();
+          }
+          return observableCombineLatest([initialManifest$,
+                                          waitForEMEReady$]).pipe(
```

Comment:
What about a function that just emits when the MediaSource is ready to be used.
It would take both the media element and the EMEManager (to which the `IEMEDisabledEvent` is also added) and just emit the MediaSource.

Something like:
```ts
function prepareMediaSource(
  mediaElement : HTMLMediaElement,
  emeEvent$ : Observable<IEMEManagerEvent|IEMEDisabledEvent>
) : Observable<string> {
  const mediaSource$ = openMediaSource(mediaElement)
    .pipe(shareReplay());
  return emeEvent$.pipe(
    // Emit when the EMEManager is ready to have segments being pushed
    mergeMap(function waitForEMEReady(evt) {
      if (evt.type === "eme-disabled") {
        return observableOf(undefined);
      }
      if (evt.type === "attached-media-keys") {
        return observableOf(undefined);
      }
      if (evt.type === "created-media-keys") {
        return mediaSource$.pipe(mergeMap(() => {
          evt.value.attachMediaKeys$.next();

          if (evt.value.keySystemOptions
                .disableMediaKeysAttachmentLock === true)
          {
            return observableOf({ isReady: true });
          }
          return EMPTY;
        }));
      }
      return EMPTY;
    }),
    exhaustMap(mediaSource$),
    take(1)
  );
}
```

Reply:
Yes it is much clearer.